Don't try to use aligned_alloc on macOS 10.14.x or earlier; it's not available
in the OS's libc++ even if the compiler supports it.
https://jira.mongodb.org/browse/CDRIVER-4609
--- src/libbson/src/bson/bson-memory.c.orig	2023-04-04 16:40:53.000000000 -0500
+++ src/libbson/src/bson/bson-memory.c	2023-04-08 23:47:14.000000000 -0500
@@ -19,6 +19,9 @@
 #include <string.h>
 #include <stdio.h>
 #include <errno.h>
+#ifdef __APPLE__
+#include <AvailabilityMacros.h>
+#endif
 
 #include "bson-atomic.h"
 #include "bson-config.h"
@@ -33,7 +36,7 @@
 // For compatibility with C standards prior to C11.
 static void *
 _aligned_alloc_impl (size_t alignment, size_t num_bytes)
-#if __STDC_VERSION__ >= 201112L && !defined(_WIN32)
+#if __STDC_VERSION__ >= 201112L && !defined(_WIN32) && (!defined(MAC_OS_X_VERSION_MIN_REQUIRED) || MAC_OS_X_VERSION_MIN_REQUIRED >= 101500)
 {
    return aligned_alloc (alignment, num_bytes);
 }
